### DISCUSSION TITLE
Mapped zod object

### DISCUSSION THREAD

**Discussion by lucioreyli** [colinhacks/zod#1997] [Q&A] [answered]
I'm trying to create a mapped zod validate object with `Record` but the result is not cool. The final result is an object mapped with the keys (right) but all values (entries) as `any` type.

![Screenshot 2023-02-06 at 20 22 54](https://user-images.githubusercontent.com/62335491/217110160-328cc342-e30f-404b-93a4-87162cfc2540.png)
![Screenshot 2023-02-06 at 20 22 37](https://user-images.githubusercontent.com/62335491/217110184-551dc962-26e2-40ae-9da3-d637d21f4e34.png)

So the result can make troubles with `TypeScript`, like this: 

![Screenshot 2023-02-06 at 20 33 01](https://user-images.githubusercontent.com/62335491/217110926-9873de37-1320-4862-8986-1736a08cdd85.png)

Because all values are typed as any

**Comment by JacobWeisenburger** [comment]
The code you are showing seems to have some other library that I'm not familiar with. Please use code blocks instead of pictures and give an example with only zod code that shows the issue you are having. Otherwise, it's really hard to understand what you are asking for. Thanks

**Comment by lucioreyli** [comment]
I basically have a trouble using Mapped type with Record (or for/in type) because all the properties is getting `any` type because the object is typed with Any, but I don't know another type to replace the `ZodTypeAny`.

```typescript
const userDTOSchema: Record<string, ZodTypeAny> = {
    name: z.string(),
    email: z.string().email(),
    password: z.string(),
}

const getUserSchema = z.object(userDTOSchema);
const { 
 name, // any
 email, // any
 password, // any
} = getUserSchema.parse(objectToValidateWithZod);
```

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const userDTOSchema = z.object( {
    name: z.string(),
    email: z.string().email(),
    password: z.string(),
} )
type UserDTO = z.infer<typeof userDTOSchema>
// type UserDTO = {
//     name: string
//     email: string
//     password: string
// }

console.log(
    userDTOSchema.parse( { name: 'John', email: 'john@doe.com', password: '123' } )
) // { name: 'John', email: 'john@doe.com', password: '123' }
```

**Reply by lucioreyli** [reply]
yes, but I don't wanna use `z.infer`, just, the reverse of this, like:
```typescript
type User = {
    name: string;
    email: string;
    password: string;
}

const userDTOSchema: Record<keyof User, ZodTypeAny> = {
    name: z.string(),
    email: z.string().email(),
    password: z.string(),
}
```

But the return of this is any for all the pairs key-value. And I don't know some type to replace the `ZodTypeAny` because it's causing  the problem.
If I replace the:
```typescript
const userDTOSchema: Record<keyof User, ZodTypeAny> = { ... };
```
with
```typescript
const userDTOSchema = { ... } as Record<keyof User, ZodTypeAny>;
```

"MAYBE" solve the problem, I don't know, but I will try.

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
type User = {
    name: string
    email: string
    password: string
}

const userDTOSchema = z.object( {
    name: z.string(),
    email: z.string().email(),
    password: z.string(),
} ) satisfies z.ZodObject<Record<keyof User, z.ZodTypeAny>>

type UserDTO = z.infer<typeof userDTOSchema>
// type UserDTO = {
//     name: string
//     email: string
//     password: string
// }

console.log(
    userDTOSchema.parse( { name: 'John', email: 'john@doe.com', password: '123' } )
) // { name: 'John', email: 'john@doe.com', password: '123' }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
